blockchain/state_manager/src/lib.rs

```diff
@@ -117,6 +131,98 @@ where
         Ok((state_root, rect_root))
     }
 
+    pub fn tipset_state(&mut self, tipset: &Tipset) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        trace!("tipSetState");
+
+        // if exists in cache return
+        if let Some(cid_pair) = self.cache.get(&tipset.key()) {
+            return Ok(cid_pair.clone());
+        }
+
+        if tipset.len() == 0 {
+            // NB: This is here because the process that executes blocks requires that the
+            // block miner reference a valid miner in the state tree. Unless we create some
+            // magical genesis miner, this won't work properly, so we short circuit here
+            // This avoids the question of 'who gets paid the genesis block reward'
+            let message_receipts = tipset
+                .blocks()
+                .first()
+                .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+            let cid_pair = (
+                tipset.parent_state().clone(),
+                message_receipts.message_receipts().clone(),
+            );
+            self.cache.insert(tipset.key().clone(), cid_pair.clone());
+            return Ok(cid_pair);
+        }
+
+        let block_headers = tipset.blocks().to_vec();
+        // generic constants are not implemented yet this is a lowcost method for now
+        let cid_pair = self.compute_tipset_state(&block_headers, &HashMap::new())?;
+        self.cache.insert(tipset.key().clone(), cid_pair.clone());
+        Ok(cid_pair)
+    }
+
+    pub fn compute_tipset_state<'a>(
+        &'a self,
+        blocks: &[BlockHeader],
+        forks_at_heights: &'a HashMap<ChainEpoch, ForkFunctions<'a, DB>>,
+    ) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        trace!("compute tipset state");
```

Comment:
I don't really think this trace log is very useful, go impl uses it for a span to track time in function, this just logs out that message

Reply:
Yeah in general wasn't sure how detailed we wanted to go with our tracing but I can expand on this for sure